### ISSUE TITLE
False Positive in "use_self"

### ISSUE THREAD

**Issue by darnir** [rust-lang-nursery/rust-clippy#2653]
I think this is the last of the issues I had with clippy this time:

The `use_self` lint seems to have a false positive when the object uses generics. Rust doesn't currently support "Self" with generics. For example, cliipy provides the following warning:

```
warning: unnecessary structure name repetition
   --> libs/mmultiset/src/lib.rs:595:35
    |
595 |     fn add_assign(&mut self, rhs: CappedMultiset<U>) {
    |                                   ^^^^^^^^^^^^^^^^^ help: use the applicable keyword: `Self`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.193/index.html#use_self
```

However, switching it to `fn add_assign(&mut self, rhs: Self<U>)` is incorrect Rust syntax.

**Comment by oli-obk** [comment]
Have you tried using Self without generic args?

**Comment by darnir** [comment]
Aah. Yes. I did not realise that I could use Self without the generics in this case.
